File: frontend/lib/models/analysis.dart
import 'track.dart';

class AnalysisResult {
  const AnalysisResult({required this.track, required this.summary, required this.chords, required this.scales, required this.modes, required this.tablature, required this.stems});

  final Track track;
  final Map<String, dynamic> summary;
  final List<Map<String, dynamic>> chords;
  final List<Map<String, dynamic>> scales;
  final List<Map<String, dynamic>> modes;
  final Map<String, dynamic> tablature;
  final Map<String, dynamic> stems;

  factory AnalysisResult.fromJson(Map<String, dynamic> json) => AnalysisResult(
        track: Track.fromJson(json['track'] as Map<String, dynamic>),
        summary: Map<String, dynamic>.from(json['summary'] as Map),
        chords: (json['chords'] as List<dynamic>).map((item) => Map<String, dynamic>.from(item as Map)).toList(),
        scales: (json['scales'] as List<dynamic>).map((item) => Map<String, dynamic>.from(item as Map)).toList(),
        modes: (json['modes'] as List<dynamic>).map((item) => Map<String, dynamic>.from(item as Map)).toList(),
        tablature: Map<String, dynamic>.from(json['tablature'] as Map),
        stems: Map<String, dynamic>.from(json['stems'] as Map),
      );
}

File: backend/app/api/routes/analysis.py
from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.orm import Session

from app.api.deps import get_current_user, get_db
from app.models import AnalysisResult, Track, User
from app.schemas import AnalysisRead, TrackRead
from app.services.separation import separate_instruments

router = APIRouter(prefix="/analysis", tags=["analysis"])


@router.post("/{track_id}/separate", response_model=AnalysisRead)
def separate_track(track_id: str, db: Session = Depends(get_db), current_user: User = Depends(get_current_user)) -> AnalysisRead:
    track = db.get(Track, track_id)
    if not track or track.owner_id != current_user.id:
        raise HTTPException(status_code=404, detail="Track not found")
    stems = separate_instruments(track.storage_path, track.id)
    if track.analysis:
        track.analysis.stems = stems
    else:
        track.analysis = AnalysisResult(track_id=track.id, summary={}, chords=[], scales=[], modes=[], tablature={}, stems=stems)
    track.status = "separated"
    db.commit()
    db.refresh(track)
    result = track.analysis
    return AnalysisRead(
        track=TrackRead.model_validate(track),
        summary=result.summary,
        chords=result.chords,
        scales=result.scales,
        modes=result.modes,
        tablature=result.tablature,
        stems=result.stems,
    )
